rfcs/0050-vm-syscalls-3/0050-vm-syscalls-3.md

```diff
@@ -12,174 +12,139 @@ Created: 2023-04-17
+The arguments used here index, source, bounds, place, argc and argv follow the usage described in [EXEC].
 
-This syscall might return the following results:
+### Pipe
+[Pipe]: #pipe
 
-- 0: Success.
-- 1-3: Reserved. These values are already assigned to other syscalls.
-- 4: Elf format error
-- 5: Exceeded max content length.
-- 6: Wrong memory limit
-- 7: Exceeded max peak memory
+This syscall create a pipe with read-write pair of file descriptions. The file descriptor with read permission is located at `fds[0]`, and the corresponding file descriptor with write permission is located at `fds[1]`.
 
-Note that now we have a new limit called *Peak Memory Usage*. The maximum memory usage of the parent script and its descendant children cannot exceed this value. Currently, this limit is set at 32M.
-
-Unlike cycles which always increase, the current memory can decrease or increase. When a child script is returned, the occupied memory is freed. This makes current memory usage lower.
+```c
+int ckb_pipe(uint64_t fds[2]);
+```
 
+File descriptors can be passed to a child process via the `inherited_fds` parameter of the Spawn syscall.
 
-### Get Memory Limit
-[Get Memory Limit]: #get-memory-limit
+### Inherited File Descriptors
+[Inherited File Descriptors]: #inherited-file-descriptors
 
-Get the maximum available memory for the current script.
+This syscall retrieves the file descriptors available to the current process, which are passed in from the parent process. These results are copied from the `inherited_fds` parameter of the Spawn syscall.
 
 ```c
-int ckb_get_memory_limit();
+int ckb_inherited_file_descriptors(uint64_t* fd, size_t* count);
 ```
 
-For the script(prime script) directly invoked by CKB, it will always return 8(4M). For the child script invoked by prime script or other child script, it depends on the parameters set by *Spawn*.
+When returning from the syscall, the syscall fills `fd` with the file descriptors in unit of `uint64_t` and fills in `count` with the count of corresponding file descriptors. The actual count of file descriptor written to `fd` is the minimum value between the count of `inherited_fds` in the Spawn syscall and the input value pointed to by `count` before syscall.
+
+By providing `NULL` as `fd` and a `size_t` pointer with a value of 0 as `count`, this syscall can be used to fetch the count of `inherited_fds` in the Spawn syscall. However, by providing `NULL` as `fd` and a non-zero value as `count`, this syscall fails.
 
-### Current Memory
-[Current Memory]: #current-memory
+### Read
+[Read]: #read
 
-Get the Current Memory Usage. The result is the sum of the memory usage of the parent script and the child script.
+This syscall reads data from a pipe via a file descriptor. The syscall Read attempts to read up to value pointed by length bytes from file descriptor fd into the buffer, and the actual length of data read is written back to the length parameter. The syscall may pause the execution of current process.
```

Comment:
I think we will need a more detailed description of read & write syscalls:

* What could contribute to the fact that less than requested bytes are read / written?
* Will the VM block so read syscalls can get more bytes to read? Will the VM block so write syscalls can write more bytes?
* And it would be best if we can describe a way to tell a channel is closed by using read / write syscall

I do recommend checking out the man page for Linux read / write syscalls.

Reply:
Add more detailed under the Write system call:

- If the Writer writes W bytes and the Reader reads R bytes where W > R, the Writer will block, and the Reader will return immediately with R bytes in `*length`. The Reader can then call `ckb_read` again to read the remaining W - R bytes.
- If the Writer writes W bytes and the Reader reads R bytes where W <= R, both the Writer and the Reader will return immediately with W bytes in `*length`.